## (작업형3) 연습문제

#### 1.주어진 데이터(Rabbit_Five.csv)는 신약 개발을 위해 실험에 사용된 데이터이다. 아래 데이터를 이용하여 실험군(MDL)과 대조군(Control) 간 혈압 변화(BPchange)가 차이가 있는 지를 쌍체표본 t-검정(paired t-test)를 통해 답하고자 한다.

##### (a) : 표본 평균을 계산하는 것으로 데이터로부터 (실험군 혈압 변화 -대조군 혈압변화)을 계산한 후 문제에서 요구하는 형태에 맞게 출력한다.

In [1]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [2]:
import pandas as pd
df1 = pd.read_csv("/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/Rabbit_Five.csv", encoding = 'cp949')

In [19]:
from scipy.stats import ttest_rel

df1_control = df1[df1['Treatment']=='Control'].reset_index()
df1_mdl = df1[df1['Treatment']=='MDL'].reset_index()
df1_con_bp = df1_control["BP_change"]
df1_mdl_bp = df1_mdl["BP_change"]
df1_mean = (df1_mdl_bp - df1_con_bp).mean()
round(df1_mean, 2)

-4.68

In [ ]:
#### 연습문제 01 Solution
import pandas as pd
exam1 = pd.read_csv('data/연습문제/Rabbit_Five.csv', encoding = 'cp949')


# 필요한 패키지, 함수 호출
from scipy.stats import ttest_rel

In [ ]:
# 필요한 컬럼 각각 할당
BP_change = exam1['BP_change']
Treatment = exam1['Treatment']

# Treatment가 Control인 경우(대조군)와 MDL인 경우(실험군)의 BP_change 값 각각 할당
BPC_Treat = BP_change[Treatment == "MDL"].reset_index(drop = True)
BPC_Control = BP_change[Treatment == "Control"].reset_index(drop = True)

## (a) 점추정량 = mean(PC_Treat - PC_Control)
diff_avg = (BPC_Treat - BPC_Control).mean()
diff_avg = round(diff_avg, 2)
print(diff_avg)

-4.68


##### (b) - (c) : ttest_rel() 함수를 통해 가설 검정을 수행하고 결과에서 제공하는 검정통계량과 p - 값을 각각 문제에서 요구하는 형태에 맞게 출력한다. p - 값이 0.001이므로 본 가설 검정은 ‘기각’한다.

In [21]:
from scipy import stats

In [38]:
t_stat, p_value = stats.ttest_rel(df1_mdl_bp, df1_con_bp)
print('t_stat: {:.3f} \np_value: {:.3f}'.format(t_stat, p_value))
if p_value < 0.05:
  result = "기각한다"
else:
  result = "기각하지 않는다"

print("p_value가 {:.3f}이므로 본 가설은 {}.".format(p_value, result))

t_stat: -3.669 
p_value: 0.001
p_value가 0.001이므로 본 가설은 기각한다.


In [ ]:
## (b)-(c)
# 대응표본 t검정 수행
a = ttest_rel(BPC_Treat, BPC_Control)

# (b) 검정통계량
stat = a.statistic
stat = round(stat, 2)
print(stat)

# (c) p-값/기각여부
pval = a.pvalue
pval = round(pval, 3)
print(pval)
print('기각')

-3.67
0.001
기각


---

#### 2.주어진 데이터(mtcars2.csv)를 통해 변속기 종류(am)에 따라 마력(hp)에 대한 분산이 차이가 있는 지를 분산비 검정(F test to compare two variances)를 통해 답하고자 한다.

##### (a) : 수동변속기(am = 1)과 자동변속기(am = 0)의 각각 마력(hp)를 산출하여 문제에서 요구하는 형태에 맞게 출력한다.

In [39]:
import pandas as pd
df2 = pd.read_csv('/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/mtcars2.csv', encoding = 'cp949')

In [40]:
df2.head()

mpg  cyl   disp   hp  drat     wt   qsec  vs  am  gear  carb
0  21.0    6  160.0  110  3.90  2.620  16.46   0   1     4     4
1  21.0    6  160.0  110  3.90  2.875  17.02   0   1     4     4
2  22.8    4  108.0   93  3.85  2.320  18.61   1   1     4     1
3  21.4    6  258.0  110  3.08  3.215  19.44   1   0     3     1
4  18.7    8    NaN  175  3.15  3.440  17.02   0   0     3     2

In [43]:
df2_am1 = df2[df2['am'] == 1]
df2_am0 = df2[df2['am'] == 0]

In [51]:
df2_am1_hp = df2_am1['hp'].reset_index(drop='True')
df2_am0_hp = df2_am0['hp'].reset_index(drop='True')

In [ ]:
#### 연습문제 02 Solution
import pandas as pd
exam2 = pd.read_csv('data/연습문제/mtcars2.csv', encoding = 'cp949')

# 필요한 패키지, 함수 호출
from scipy.stats import f

# 필요한 컬럼 각각 할당
am = exam2['am']
hp = exam2['hp']

# 수동변속기(am = 1)인 자동차의 마력(hp), 자동변속기(am = 0)인 자동차의 마력(hp) 각각 할당
hp_manual = hp[am ==1].reset_index(drop = True)
hp_auto = hp[am == 0].reset_index(drop = True)

## (a) 수동변속 표본분산/자동변속 표본분산
var_ratio = hp_manual.var()/hp_auto.var()
print(round(var_ratio,2))

2.43


##### (b)-(c) : f 클래스로 분포 객체를 생성 한 후 해당 객체의 메소드 cdf()를 이용해 p값을 각 문제에서 요구하는 형태에 맞게 출력한다. p - 값이 0.043이므로 본 가설 검정은 ‘기각’한다.

In [60]:
import pandas as pd
from scipy import stats

# 변속기 종류(am)에 따른 그룹별 마력(hp) 데이터 분리
hp_manual = df2[df2['am'] == 1]['hp']
hp_auto = df2[df2['am'] == 0]['hp']

# 분산비 검정 수행
f_stat, p_value = stats.levene(hp_manual, hp_auto)

# 결과 출력
print('F-statistic: {:.3f} \np-value: {:.3f}'.format(f_stat, p_value))

# p-value 해석
if p_value < 0.05:
    result = "기각"
else:
    result = "기각하지 않는다"

print("p-value가 {:.3f}이므로 본 가설은 {}.".format(p_value, result))


F-statistic: 0.184 
p-value: 0.671
p-value가 0.671이므로 본 가설은 기각하지 않는다.


In [56]:
from scipy import stats

f_stat, p_value = stats.levene(df2_am1_hp, df2_am0_hp)
print('f_stat:{:.3f} \np_value: {:.3f}'.format(f_stat, p_value))

f_stat:0.184 
p_value: 0.671


In [ ]:
## (b)-(c)
# F검정 수행
# (b) 검정통계량 = (a)와 같음
stat = var_ratio
print(stat)

# (c) 유의확률
# 자유도
df1, df2 = len(hp_manual) - 1, len(hp_auto) - 1

# F 분포로 확률 계산
pval = 1 - f.cdf(stat, dfn = df1, dfd = df2) # Pr[F > stat]

# 정답 출력을 위해 반올림
a = round(var_ratio, 2)
b = round(stat, 2)
c = round(pval, 3)

# 정답 출력
print(a) # (a)
print(b) # (b) 검정통계량
print(c) # (c) p - 값
print('기각') # 기각여부

2.431605932653339
2.43
2.43
0.043
기각


---

#### 3. 주어진 데이터(고객_등급리스트.csv)를 통해 고객군(Segment)과 지역(Region)간 관련이 있는지를 독립성 검정(Test of independence)을 통해 답하고자 한다.

##### (a) : segment와 region의 교차표를 생성하고 chi2_contigency() 함수를 통해 가설검정을 수행한 후 E23 기대도수를 문제에서 요구하는 형태에 맞게 출력한다.

In [61]:
import pandas as pd
from scipy import stats
df3 = pd.read_csv('/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/고객_등급리스트.csv', encoding = 'cp949')
df3.head()

Customer_ID     ID             Name Age             City           State  \
0          NaN  12520      Claire Gute  6*        Henderson        Kentucky   
1          NaN  13045  Darrin Van Huff  3*      Los Angeles      California   
2          NaN  20335   Sean O'Donnell  6*  Fort Lauderdale         Florida   
3          NaN  11710  Brosina Hoffman  2*      Los Angeles      California   
4          NaN  10480     Andrew Allen  5*          Concord  North Carolina   

  Region    Segment  
0  South   Consumer  
1   West  Corporate  
2  South   Consumer  
3   West   Consumer  
4  South   Consumer

In [86]:
from scipy.stats import chi2_contingency

contingency_table = pd.crosstab(df3['Segment'], df3['Region'])

chi2, p, dof, expected = chi2_contingency(contingency_table)
print(chi2, p)

9.488312197551783 0.1479205256846858


In [82]:
from scipy.stats import chi2_contingency

#교차표 생성
contingency_table = pd.crosstab(df3['Segment'], df3['Region'])

chi2, p, dof, expected = chi2_contingency(contingency_table)

print('{:.3f}\n{:.3f}'.format(chi2, p))

9.488
0.148


In [ ]:
#### 연습문제 03 Solution
import pandas as pd
exam3 = pd.read_csv('data/연습문제/고객_등급리스트.csv', encoding = 'cp949')

# 필요한 패키지, 함수 호출
from scipy.stats import chi2_contingency
import numpy as np

# 교차표 생성
tb = pd.crosstab(exam3['Segment'], exam3['Region'])

# 카이제곱 검정 수행
# chi2_contingency의 결과는 카이제곱통계량, 유의확률, 자유도, 기대도수를 반환함
chi2, pval, df, expected = chi2_contingency(tb)

# (a)E23 : expected의 (1,2) 인덱스 번호 추출
e23 = expected[1,2]
e23 = round(e23, 2)
print(e23)

15.74


##### (b) - (c) : 위의 chi2_contingency 결과를 통해 검정통계량과 p - 값을 각각 문제에서 요구하는 형태에 맞게 출력한다. p - 값이 0.148이므로 본 가설 검정은 ‘채택’한다.

In [ ]:
# (b) 검정통계량
chi2 = chi2.astype('int') # 정수 변환
print(chi2)

# (c) p값/기각 여부
pval = round(pval, 3)
print(pval)
print('채택')

9
0.148
채택


---

#### 4. 주어진 데이터(Cars93.csv)를 통해 가격(Price)이 정규 분포를 따르는 지를 샤피로- 윌크(Shapiro Wilk Test)를 통해 답하고자 한다. 가설은 아래와 같다.(단, 결측치는 무시할 것)

##### (a) : dropna()를 통해 결측치를 제거한 뒤, Price의 표본평균을 mean()을 사용해 계산하고 문제에서 요구하는 형태에 맞게 출력한다.

In [90]:
import pandas as pd
df4 = pd.read_csv('/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/Cars93.csv', encoding = 'cp949')
df4.head()

Manufacturer    Model     Type  Min_Price  Price  Max_Price  MPG_city  \
0        Acura  Integra    Small       12.9   15.9       18.8        25   
1        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2         Audi       90  Compact       25.9   29.1       32.3        20   
3         Audi      100  Midsize       30.8    NaN       44.6        19   
4          BMW     535i  Midsize       23.7    NaN       36.2        22   

   MPG_highway             AirBags DriveTrain  ... Length  Wheelbase  Width  \
0           31                 NaN      Front  ...    177        102     68   
1           25  Driver & Passenger      Front  ...    195        115     71   
2           26         Driver only      Front  ...    180        102     67   
3           26  Driver & Passenger      Front  ...    193        106     70   
4           30         Driver only       Rear  ...    186        109     69   

   Turn_circle  Rear_seat_room Luggage_room  Weight   Origin  Unnamed: 26  \
0         37.0            26.5         11.0    2705  non-USA            0   
1         38.0            30.0         15.0    3560  non-USA            1   
2         37.0            28.0         14.0    3375  non-USA            2   
3          NaN            31.0         17.0    3405  non-USA            3   
4          NaN            27.0         13.0    3640  non-USA            4   

            Make  
0  Acura Integra  
1   Acura Legend  
2        Audi 90  
3       Audi 100  
4       BMW 535i  

[5 rows x 28 columns]

In [99]:
df4_na = df4['Price'].dropna()

In [101]:
df4_na

0     15.9
1     33.9
2     29.1
5     15.7
6     20.8
      ... 
88    19.7
89    20.0
90    23.3
91    22.7
92    26.7
Name: Price, Length: 84, dtype: float64

In [107]:
from scipy.stats import shapiro
stat, p_value = shapiro(df4_na)
print('{:.3f}  {:.3f}'.format(stat, p_value))

0.855  0.000


In [104]:
# 가격(Price) 데이터에서 결측치 제거
price_data = df4['Price'].dropna()

# 샤피로-윌크 검정 수행
stat, p_value = shapiro(price_data)

# 결과 출력
print('Shapiro-Wilk test statistic: {:.3f}'.format(stat))
print('p-value: {:.3f}'.format(p_value))

Shapiro-Wilk test statistic: 0.855
p-value: 0.000


In [ ]:
#### 연습문제 04 Solution
import pandas as pd
exam4 = pd.read_csv('data/연습문제/Cars93.csv', encoding = 'cp949')

# 필요한 패키지, 함수 호출

from scipy.stats import shapiro
# 필요한 컬럼 할당
Price = exam4['Price'].copy().dropna()

# (a) 표본평균
avg = Price.mean()
avg = round(avg, 2)
print(avg)

19.05


##### (b) - (c) : shapiro() 함수를 통해 가설검정을 수행하고 결과에서 제공하는 검정통계량과 p - 값을 각각 문제에서 요구하는 형태에 맞게 출력한다. p - 값이 0이므로 본 가설 검정은 ‘기각’한다.

In [ ]:
# 샤피로 윌크 검정 수행
stat, pval = shapiro(Price)

# (b) 검정통계량
stat = round(stat, 2)
print(stat)

# (c) p값/기각 여부
pval = round(pval, 4)
pval = int(pval)
print(pval)
print('기각')

0.85
0
기각


---

#### 5. 주어진 데이터(Cars93.csv)를 통해 마일당엔진회전수(Rev_per_mile)과 마력(Horsepower)과의 상관관계를 알아보고 상관계수의 유의성을 피어슨 상관계수 검정(Pearson's Correlation Coefficient Test)를 통해 답하고자 한다. 가설은 아래와 같다.

##### (a) : pearsonr() 함수를 통해 ρˆ 를 구하고 문제에서 요구하는 형태에 맞게 출력한다.

In [108]:
import pandas as pd
import numpy as np
df5 = pd.read_csv('/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/Cars93.csv')
df5.head()

Manufacturer    Model     Type  Min_Price  Price  Max_Price  MPG_city  \
0        Acura  Integra    Small       12.9   15.9       18.8        25   
1        Acura   Legend  Midsize       29.2   33.9       38.7        18   
2         Audi       90  Compact       25.9   29.1       32.3        20   
3         Audi      100  Midsize       30.8    NaN       44.6        19   
4          BMW     535i  Midsize       23.7    NaN       36.2        22   

   MPG_highway             AirBags DriveTrain  ... Length  Wheelbase  Width  \
0           31                 NaN      Front  ...    177        102     68   
1           25  Driver & Passenger      Front  ...    195        115     71   
2           26         Driver only      Front  ...    180        102     67   
3           26  Driver & Passenger      Front  ...    193        106     70   
4           30         Driver only       Rear  ...    186        109     69   

   Turn_circle  Rear_seat_room Luggage_room  Weight   Origin  Unnamed: 26  \
0         37.0            26.5         11.0    2705  non-USA            0   
1         38.0            30.0         15.0    3560  non-USA            1   
2         37.0            28.0         14.0    3375  non-USA            2   
3          NaN            31.0         17.0    3405  non-USA            3   
4          NaN            27.0         13.0    3640  non-USA            4   

            Make  
0  Acura Integra  
1   Acura Legend  
2        Audi 90  
3       Audi 100  
4       BMW 535i  

[5 rows x 28 columns]

In [110]:
df5_mile = df5['Rev_per_mile']
df5_holse = df5['Horsepower']

from scipy.stats import pearsonr

corr_coef, p_value = pearsonr(df5_mile, df5_holse)
print('{:.3f}\n{:.3f}'.format(corr_coef, p_value))

-0.502
0.000


In [ ]:
#### 연습문제 05 Solution
import pandas as pd
import numpy as np
exam5 = pd.read_csv('data/연습문제/Cars93.csv')

# 상관분석에 필요한 컬럼명 저장
Horsepower = exam5['Horsepower']
RPM = exam5['Rev_per_mile']

# 필요한 패키지, 함수 호출
from scipy.stats import pearsonr

# 상관계수 검정 수행
rho, pval = pearsonr(Horsepower, RPM)

# (a) 표본상관계수
rho = round(rho, 3)
print(rho)

-0.502


##### (b) - (c) : pearsonr() 함수를 통해 구한 ρˆ 를 이용해 검정통계량을 구하고, p - 값을 각 문제에서 요구하는 형태에 맞게 출력한다. p - 값이 0이므로 본 가설 검정은 ‘기각’한다.

In [ ]:
# (b) 검정통계량
stat = rho/np.sqrt( (1-rho**2)/(len(Horsepower) - 2) )
stat = round(stat, 2)
print(stat)

# (c) p값/기각 여부
pval = int(pval)
print(pval)
print('기각')

-5.54
0
기각


---

#### 6. 제공된 데이터(USArrests.csv)는 미국 50개 주의 범죄와 체포와 관련된 데이터이다. 차원 축소를 주성분 분석을 통해 수행하고자 한다.

##### (a) : PCA() 클래스로 주성분 객체 생성 후 메소드 fit_transform를 통해 데이터를 적합한 후 주성분분석을 수행한다. 이후 적합된 주성분객체로 부터 components_.[]를 이용해 첫번째 주성분에서 폭력범죄가 기여하는 가중치를 문제에서 요구하는 형태에 맞게 출력한다.

In [111]:
import pandas as pd
df6 = pd.read_csv('/content/drive/MyDrive/빅분기 실기 2024/data/연습문제/USArrests.csv')
df6.head()

Murder  Assault  UrbanPop  Rape
0    13.2      236        58  21.2
1    10.0      263        48  44.5
2     8.1      294        80  31.0
3     8.8      190        50  19.5
4     9.0      276        91  40.6

In [113]:
from sklearn.decomposition import PCA

pca = PCA(n_components = 4)
df6_pca = pca.fit_transform(df6)

In [ ]:
#### 연습문제 06 Solution
import pandas as pd
exam6 = pd.read_csv('data/연습문제/USArrests.csv')

# 필요한 패키지, 클래스 호출
from sklearn.decomposition import PCA

# PCA 수행
pca = PCA(n_components = 4) # 주성분객체 생성
pca.fit_transform(exam6)

# (a) 첫번째 주성분의 폭력범죄 기여 가중치
# pca.components_.T에서 행은 기존 컬럼(Merder, Assault, UrbanPop, Rape)
# 열은 1~4 주성분임
weight = pca.components_.T[1, 0]
weight = round(weight, 3)
print(weight)

0.995


##### (b) - (c) : 주성분점수는 적합시에만 출력이 가능해 다시 fit_transform()로 재적합한 후 34번째 도시의 1주성분의 주성분 점수를 구하고, 적합된 객체로부터 explained_variance_ratio_를 통해 누적기여율을 구하고 각 문제에서 요구하는 형태에 맞게 출력한다.

In [ ]:
# (b) 34번째 도시의 1주성분의 주성분 점수
score = pca.fit_transform(exam6)[33,0]
score = round(score, 3)
print(score)

# (c)
# 주성분별 설명되는 분산 비율을 시리즈 객체로 저장
var_ratio = pd.Series(pca.explained_variance_ratio_)
result = round(var_ratio[0], 2)
print(result)
print(3)

-127.496
0.97
3


---

#### 7. 주어진 데이터(Cars93.csv)를 통해 마일당엔진회전수(Rev_per_mile), 중량(Weight), 길이(Length), 엔진크기(EngineSize)를 입력하면 중간가격(Price)을 예측하는 다중 선형 회귀 분석을 하고자 한다.(단, 해당 컬럼들의 결측치는 제거함)

##### (a) : 주어진 컬럼들의 결측치를 제거하고, OLS 클래스로 OLS 객체를 생성한 후 fit을 통해 모델을 적합 후 적합된 모형으로부터 summary를 통해 결과를 확인하고 출력되는 결과에서 결정계수를 찾아 문제에서 요구한 형태에 맞게 출력한다.

In [ ]:
#### 연습문제 07 Solution
import pandas as pd
exam7 = pd.read_csv('data/연습문제/Cars93.csv')

# 필요한 패키지, 클래스 호출
import statsmodels.api as sm

# 회귀 분석에 필요한컬럼 별도 지정
colnm = ['Price', 'Rev_per_mile', 'Weight','Length', 'EngineSize']
samp = exam7[colnm].dropna() # 결측치 제거

#회귀분석 수행
# y, X에 각각 할당
y = samp['Price']
X = samp[['Rev_per_mile', 'Weight','Length', 'EngineSize']]
X = sm.add_constant(X) # 절편항 적합을 위해 상수벡터 추가

# 모델 적합
model = sm.OLS(y,X) # OLS 객체 생성
result = model.fit() # fit 메소드를 통해 모형 적합

# (참고)
# result.params # 회귀 계수만 추출
# result.tvalues # t통계량만 추출
result.summary() # 해당 코드를 통해 회귀분석 통합 결과를 확인하고 값을 입력하면 됨
# 결과 출력
# (a) 결정계수
r_square = 0.396
print(r_square)

0.396


##### (b) - (d) : summary()의 출력 내용에서 Weight의 추정 회귀계수와, Weight의 P > |t|, Weight의 회귀 계수에 대한 95%의 신뢰구간을 각각 문제에서 요구하는 형태에 맞게 출력한다.

In [ ]:
# (b) 문제의 의도는
# Weight의 추정 회귀 계수를 출력하는 것이다.
b = 0.0023
print(b)

# (c) 문제의 의도는
# Weight의 P > |t|을 통해 회귀 계수를 검정하는 것이다.
pval = 0.158
print(pval)

# (d) 문제의 의도는
# Weight의 회귀 계수에 대한 95% 신뢰구간을 구하는 것이다. 아래의 함수 결과를 통해 상한 값을 입력하면 됨
# result.conf_int(alpha = 0.05, cols = None)
upper = 0.005406
upper = round(upper,4)
print(upper)

0.0023
0.158
0.0054


---

#### 8. 제공된 데이터(job.csv)는 취업 현황 분석을 위해 사용된 데이터의 일부이다. 여러 특성(x1, x2, x3)를 통해 취업 성공(y) 여부를 예측하는 로지스틱 회귀 분석을 하고자 한다. x2 컬럼은 성별에 대한 정보로, 모형 적합시 남성(M)을 1로, 여성(F)로 인코딩하여 분석하되, 별도의 규제화는 진행하지 않음)

##### (a) : x2 컬럼을 인코딩한 후 GLM 객체를 생성하고 모형을 적합한다. 이 후, summary를 통해 결과를 출력하고 해당 결과로부터 절편항 추정 회귀 계수를 찾아 문제에서 요구하는 형태에 맞게 출력한다.

In [ ]:
#### 연습문제 08 Solution
import pandas as pd
exam8 = pd.read_csv('data/연습문제/job.csv')

# 필요한 패키지, 클래스 호출
import statsmodels.api as sm
import numpy as np

# x2 컬럼 : M -> 1, F -> 0
exam8['x2'] = exam8['x2'].map({'M' : 1, 'F' : 0})

# y, X에 각각 할당
y = exam8['y']
X = exam8[['x1', 'x2', 'x3']]
X = sm.add_constant(X) # 절편항 적합을 위해 상수벡터 추가

# 모델 적합
model = sm.GLM(y, X, family = sm.families.Binomial())
result = model.fit()

# 결과 출력
# (a) 절편항 추정 회귀 계수
b0 = -0.808 # result.summary() 로 확인 가능
print(b0)

-0.808


##### (b) - (c) : np.exp() 함수를 통해 여성에 비해 남성의 성공에 대한 오즈가 몇배 감소하는지(남성의 성공 오즈/여성의 성공 오즈)를 구하고, 적합된 로지스틱모형 객체로부터 predict_proba()를 통해 9번째 사람의 성공 예측확률을 인덱싱하여 문제에서 요구하는 형태에 맞게 출력한다.

In [ ]:
# (b) 여성에 비해 남성의 성공에 대한 오즈가 몇 배인지를 구하려면
# 오즈비 = 남성의 성공 오즈/여성의 성공 오즈
# = x2 컬럼이 성별이므로 exp(beta2)를 구하면 됨
odds_ratio = round(np.exp(-0.1575), 3) # result.summary() 로 확인 가능
print(odds_ratio)

# (c) 9번 째 사람의 성공 예측 확률
y_prob = round(result.predict(X)[8], 4)
print(y_prob)
print(0)

0.854
0.5344
0


---

#### 9. 주어진 데이터(영화_순위리스트.csv)를 통해 장르별 예산의 평균에 차이가 있는 지를 분산분석(ANOVA)를 수행하기 전 등분산 검정(Homogeneity of Variance)인 Bartlett's Test을 수행고자 한다. 가설은 아래와 같다.

##### (a) : 장르별 예산 값을 할당하고, 집단별 표본 분산과 관측치 수를 구한 후 log(합동분산)을 문제에서 요구하는 형태에 맞게 출력한다.

In [ ]:
#### 연습문제 09 Solution
import pandas as pd
exam9 = pd.read_csv('data/연습문제/영화_순위리스트.csv', encoding = 'cp949')

# 필요한 패키지, 함수 호출
from scipy.stats import bartlett
import numpy as np

# 필요한 컬럼 각각 할당
genre = exam9['장르']
budget = exam9['예산']

# 장르별 예산 값 할당
budget_thriller = budget[genre == 'Thriller']
budget_comedy = budget[genre == 'Comedy']
budget_drama = budget[genre == 'Drama']
budget_action = budget[genre == 'Action']

## (a) 합동분산(pooled variancer)
# 집단별 표본 분산
var_i = [budget_thriller.var(), budget_comedy.var(), budget_drama.var(), budget_action.var()]

# 집단별 관측치 수
n_i = [len(budget_thriller), len(budget_comedy), len(budget_drama), len(budget_action)]

# log(합동분산) 계산
N = sum(n_i)
k = 4 # 집단의 수
log_sp2 = np.log(sum(np.subtract(n_i, 1) * var_i)/(N-k))
log_sp2 = round(log_sp2, 3)
print(log_sp2)

16.542


##### (b) - (c) : bartlett 함수를 통해 Bartlett 검정을 수행하고, 검정통계량과 p - 값을 각 문제에서 요구한 형태에 맞게 출력한다. p - 값이 0이므로 본 가설 검정은 ‘기각’한다.

In [ ]:
## (b)-(c)
# Bartlett Test 수행
stat, pval = bartlett(budget_thriller, budget_comedy, budget_drama, budget_action)

# (b) 검정통계량
stat = round(stat, 2)
print(stat)

# (c) p-값/기각여부
pval = round(pval, 4)
print(pval)
print('기각')

13.44
0.0038
기각


# (끝)